man/hfst-pmatch.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH HFST-PMATCH "1" "April 2026" "HFST" "User Commands"
.SH NAME
hfst-pmatch \- =perform matching/lookup on text streams
.SH SYNOPSIS
.B hfst-pmatch
[\fI\,OPTIONS\/\fR...] \fI\,TRANSDUCER\/\fR
.SH DESCRIPTION
perform matching/lookup on text streams
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.TP
\fB\-\-colour\fR[=\fI\,WHEN\/\fR]
Print in colour WHEN:
.TP
\fB\-\-color\fR[=\fI\,WHEN\/\fR]
always, never, auto (default)
.SS "Input/Output options:"
.TP
\fB\-i\fR, \fB\-\-input\fR=\fI\,INFILE\/\fR
Read input transducer from INFILE
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,OUTFILE\/\fR
Write output transducer to OUTFILE
.SS "Pmatch options:"
.TP
\fB\-n\fR  \fB\-\-newline\fR
Newline as input separator (default is blank line)
.TP
\fB\-x\fR  \fB\-\-extract\-patterns\fR
Only print tagged parts in output
.TP
\fB\-l\fR  \fB\-\-locate\fR
Only print locations of matches
.TP
\fB\-w\fR  \fB\-\-print\-weights\fR
In locate mode, include weights of the matches
.TP
\fB\-c\fR  \fB\-\-count\-patterns\fR
Print the total number of matches when done
.TP
\fB\-\-delete\-patterns\fR
Replace matches with opening tags
.TP
\fB\-\-no\-mark\-patterns\fR
Don't tag matched patterns
.TP
\fB\-\-max\-context\fR
Upper limit to context length allowed
.TP
\fB\-\-max\-recursion\fR
Upper limit for recursion
.TP
\fB\-\-weight\-cutoff\fR=\fI\,W\/\fR
Upper limit for allowed weight
.TP
\fB\-t\fR, \fB\-\-time\-cutoff\fR=\fI\,S\/\fR
Limit search after having used S seconds per input
.TP
\fB\-p\fR  \fB\-\-profile\fR
Produce profiling data
.PP
Use standard streams for input and output.
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://github.com/hfst/hfst/issues>
.PP
hfst\-pmatch home page:
<https://github.com/hfst/hfst/wiki/HfstPmatch>
.br
General help using HFST software:
<https://github.com/hfst/hfst/wiki>
.SH COPYRIGHT
Copyright \(co 2017 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS for hfst-tools and library

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

## Noteworthy changes in 3.17.1

* fixes logic error where `hfst-lexc` would treat a warning about implicit
  alphabet in XREs as a critical error even when `-Werror` is not asked

## Noteworthy changes in 3.17.0

* API breaks: HfstTokenizer, LexcCompiler, `read_in_att_format` functions
  in HfstTransducer, HfstBasicTransducer, etc. to accommodate the following
  unicode handling changes and error reporting:
* Character "tokenisers" that used to operate on bytes or UTF-8 sequences will
  now operate on [graphemes](https://www.unicode.org/reports/tr29/tr29-23.html#Grapheme_Cluster_Boundaries),
  this means that by default arcs in FSAs are expected to match combining
  grapheme clusters, e.g. a base character plus all diacritics.
  The implementation is based on ICU implementation of `u_brk`
    * in lexc string entries, graphemes are compiled into a arcs
    * in xerox regex curly braces, as well
    * in tokenise without --multichar mode, input is tokenised by graphemens
    * etc. etc. tbd
* negative weights yield warnings now unless turned off
* tools will have more gcc-style warning control: -Wall -Werror -Wno-this-error
* lexc warnings are systematically treated as errors with -Werror now
* colourised more warnings and errors

## Noteworthy changes in 3.16.2

* Support for `>OpenFst-1.8.0`
    * only functional change is removal of n-best paths search for log weight
      automata, it probably has errored out before but has slightly different
      error now (HfstException instead of openfst exit)
* Optimisations to hfst-pmatch2fst alphabet handling

## Noteworthy changes in 3.16.1

* Bug fixes:

    * hfst-tokenize escapes both \r and \n now

    * regexp parsing more robust with comments and linebreaks

* Removal of some bundled libraries including foma and openfst, install
  suitable versions to compile support for them

* ICU stabilised as unicode handler over all

* experimental C API for rust bindings (unstable)

* new tools (testing):

    * hfst-check-alpha for checking and comparing alphabets in automata and
      their metadata

    * hfst-dump-alphabets for reviewing alphabets easier than hfst-summarize
      grepping

    * hfst-kill-paths for efficient path removal

* introduced clang-format to work towards standard code formatting

Noteworthy changes in 3.16.0
----------------------------

* Bugfixes:

  * During conversion between optimized-lookup and other formats, the alphabet
    was not being preserved, leading to incorrect behaviour in some cases.
    This has been fixed.

  * Fix segfaults in hfst-summarize, hfst-disjunct and foma back-end.

  * Force ICU, fixes #501.

  * Use unique_ptr for input/ouput stream memory allocation (#547).


* Improvements:

  * hfst-pmatch archives now no longer include or parse redundant alphabets,
    which in some cases can be a meaningful size and speed saving.

  * Multiple memory leaks fixed.

  * Drop support for TR1 standard assuming that C++14 is minimal requirement.

  * Drop legacy wrappers for foma and lexc (#532).

  * Use upstream Foma and by default upstream OpenFST.

  * Use C++17, or C++14 with bundled OpenFST.

  * Update external links for sfst, foma and flex.

  * Add lex/yacc sources to dist (see issue #524).

* Features:

  * A new variable in the hfst-pmatch language, "unicode-character-classes",
    which causes the use of built-in character sets to be interpreted at
    runtime using a unicode library.

  * Verbose output from hfst-pmatch2fst has been much improved, providing eg.
    a call graph of the process. It also warns about use of pmatch features
    that is likely to be incorrect and lead to errors.


Noteworthy changes in 3.15.5
----------------------------

* Fixes to pmatch:

  * Modify tracking of entry points to allow for RTNs in Captures()

  * Fix off-by-one bug in checking captures

* Fixes to other issues: #508, #488, #486, #452, #491, #494

* Fix memory leaks


Noteworthy changes in 3.15.4
----------------------------

* Import Foma patches with license change to Apache-2.0 and minimal other fixes

* Allow regeneration of Foma files so the generated files can be omitted

* Explicit license file for SFST


Noteworthy changes in 3.15.3
----------------------------

* Numerous improvements to command line tools, e.g. substitute, insert-freely, proc, fst2txt, expand


Noteworthy changes in 3.15.2
----------------------------

* Fix off-by-one bug in checking captures in pmatch


Noteworthy changes in 3.15.1
----------------------------

* Fix issue #443

* Fix issue #429

* Fix issue #409


Noteworthy changes in 3.15.0
----------------------------

* Numerous improvements in pmatch processing:

  * hfst-pmatch2fst now implements eg @include"filename.txt" as a preprocessing stage
  
  * hfst-pmatch2fst now implements Explode() and Implode() in full generality
  
  * hfst-pmatch now allows printing weights (in location mode, with --locate)
  
  * hfst.PmatchContainer.locate offered via python bindings
  
  * bugfixes and improvements hfst.PmatchContainer.tokenize

* Add tool hfst-eliminate-flags.

* Allow both escaped and unescaped zeros in multicharacters in all lexc lexicon entries (fixes issue #274).

* Also reweight state that is both initial and final in hfst-reweight (fixes issue #368).

* Do not let distributed files depend on configuration. Include man files in dist.

* Print warnings and error messages also in earlier stages of twol compilation.

* Add option --execute-and-quit (-E) to hfst-xfst.


Noteworthy changes in 3.14.0
----------------------------

* Numerous improvements to pmatching and tokenization:

  * pmatch now supports the default symbol

  * pmatch now supports reading word embeddings in the binary format


tools/src/hfst-pmatch.cc
//! @file hfst-pmatch.cc
//!
//! @brief Pmatch utility for continuous matching on input streams
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif


#include <iostream>
#include <fstream>

#include <vector>
#include <map>
#include <string>

using std::string;
using std::vector;
using std::pair;

#include <cstdio>
#include <cstdlib>
#include <cstring>

#ifdef _MSC_VER
#  include "hfst-getopt.h"
#  include "hfst-string-conversions.h"
#else
#  include <getopt.h>
#endif

#ifdef HAVE_READLINE
#include <unistd.h>
#include <readline/readline.h>
#include <readline/history.h>

/* Read a string, and point buffer to it. */
void libreadline_getline(char ** buffer)
{
  /* If the buffer has already been allocated,
     return the memory to the free pool. */
  if (*buffer)
    {
        free (*buffer);
        *buffer = (char *)NULL;
    }

  /* Get a line from the user. */
  *buffer = readline("");

  /* If the line has any text in it,
     save it on the history. */
  if (*buffer && **buffer)
    add_history (*buffer);
}

#endif

#include <math.h>
#include <errno.h>

#include "HfstExceptionDefs.h"
#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "hfst-tool-metadata.h"
#include "implementations/optimized-lookup/pmatch.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

static bool blankline_separated = true;

enum var_val{on, off, not_defined};
static var_val count_patterns = not_defined;
static var_val delete_patterns = not_defined;
static var_val extract_patterns = not_defined;
static var_val locate_mode = not_defined;
static var_val print_weights = not_defined;
static var_val mark_patterns = not_defined;
static int max_recursion = -1;
static int max_context = -1;

static double time_cutoff = 0.0;
static hfst_ol::Weight weight_cutoff = hfst_ol::INFINITE_WEIGHT;
static bool profile = false;

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] TRANSDUCER\n"
            "perform matching/lookup on text streams\n"
            "\n", program_name);
    print_common_program_options(message_out);
    print_common_unary_program_options(message_out);
    fprintf(message_out,
            "Pmatch options:\n"
            "  -n  --newline           Newline as input separator (default is blank line)\n"
            "  -x  --extract-patterns  Only print tagged parts in output\n"
            "  -l  --locate            Only print locations of matches\n"
            "  -w  --print-weights     In locate mode, include weights of the matches\n"
            "  -c  --count-patterns    Print the total number of matches when done\n"
            "      --delete-patterns   Replace matches with opening tags\n"
            "      --no-mark-patterns  Don't tag matched patterns\n"
            "      --max-context       Upper limit to context length allowed\n"
            "      --max-recursion     Upper limit for recursion\n"
            "      --weight-cutoff=W   Upper limit for allowed weight\n"
            "  -t, --time-cutoff=S     Limit search after having used S seconds per input\n"
            "  -p  --profile           Produce profiling data\n");
    fprintf(message_out,
            "Use standard streams for input and output.\n"
            "\n"
        );
    
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
    fprintf(message_out, "\n");
}

void match_and_print(hfst_ol::PmatchContainer & container,
                std::ostream & outstream,
                std::string & input_text)
{
    if (input_text.size() > 0 && input_text.at(input_text.size() - 1) == '\n') {
        // Remove final newline
        input_text.erase(input_text.size() -1, 1);
    }
    if (!container.is_in_locate_mode()) {
#ifndef _MSC_VER
        outstream << container.match(input_text, time_cutoff, weight_cutoff);
#else
        hfst::hfst_fprintf_console(stdout, "%s", container.match(input_text, time_cutoff, weight_cutoff).c_str());
#endif
        outstream << std::endl;
        if (blankline_separated) {
            outstream << std::endl;
        }
    } else {
        hfst_ol::LocationVectorVector locations = container.locate(input_text, time_cutoff, weight_cutoff);
        bool printed_something = false;
        for(const auto & location : locations) {
            if (location.at(0).output.compare("@_NONMATCHING_@") != 0) {
                printed_something = true;
#ifndef _MSC_VER
                outstream << location.at(0).start << "|" << location.at(0).length << "|"
                          << location.at(0).output << "|" << location.at(0).tag;
                if (print_weights) {
                    outstream << "|" << location.at(0).weight;
                }
                outstream << std::endl;
#else
                if (print_weights){
                    hfst::hfst_fprintf_console(stdout, "%i|%i|%s|%s|%f\n", it->at(0).start, it->at(0).length, it->at(0).output.c_str(), it->at(0).tag.c_str(), it->at(0).weight);
                } else {
                    hfst::hfst_fprintf_console(stdout, "%i|%i|%s|%s\n", it->at(0).start, it->at(0).length, it->at(0).output.c_str(), it->at(0).tag.c_str());
                }
#endif
            }
        }
        if (printed_something) {
            outstream << std::endl;
        }
    }
}


int process_input(hfst_ol::PmatchContainer & container,
                  std::ostream & outstream)
{
    std::string input_text;
    char * line = NULL;
    size_t len = 0;
    while (true) {
#ifndef _MSC_VER
#ifdef HAVE_READLINE
        if (isatty(STDIN_FILENO)) {
            libreadline_getline(&line);
            if (!line)
                break;
        } else {
            if (!(hfst_getline(&line, &len, stdin) > 0))
                break;
        }
#else
        if (!(hfst_getline(&line, &len, stdin) > 0))
            break;
#endif
#else
        std::string linestr("");
        size_t bufsize = 1000;
        if (! hfst::get_line_from_console(linestr, bufsize, true /* keep newlines */))
            break;
        line = strdup(linestr.c_str());
#endif
      
        if (!blankline_separated) {
            // newline separated
#ifndef _MSC_VER
#ifdef HAVE_READLINE
            input_text = line;
            match_and_print(container, outstream, input_text);
        } else if (line[0] == '\0' || line[0] == '\n') {
            match_and_print(container, outstream, input_text);
            input_text.clear();
        } else {
            input_text.append(line);
            if (isatty(STDIN_FILENO)) {
                input_text.push_back('\n');
            }
        }
#else
            input_text = line;
            match_and_print(container, outstream, input_text);
        } else if (line[0] == '\n') {
            match_and_print(container, outstream, input_text);
            input_text.clear();
        } else {
            input_text.append(line);
        }
#endif
#else
            input_text = line;
            match_and_print(container, outstream, input_text);
        } else if (line[0] == '\n') {
            match_and_print(container, outstream, input_text);
            input_text.clear();
        } else {
            input_text.append(line);
        }
#endif

        free(line);
        line = NULL;
    }

    if (blankline_separated && !input_text.empty()) {
        match_and_print(container, outstream, input_text);
    }
    if (count_patterns == on) {
        outstream << "\n" << container.get_pattern_count_info() << "\n";
    }
    if (profile) {
        outstream << "\n" << container.get_profiling_info() << "\n";
    }
    return EXIT_SUCCESS;
}


int parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
            {
                HFST_GETOPT_COMMON_LONG,
                HFST_GETOPT_UNARY_LONG,
                {"newline", no_argument, 0, 'n'},
                {"extract-patterns", no_argument, 0, 'x'},
                {"locate", no_argument, 0, 'l'},
                {"print-weights", no_argument, 0, 'w'},
                {"count-patterns", no_argument, 0, 'c'},
                {"delete-patterns", no_argument, 0, 'z'},
                {"no-mark-patterns", no_argument, 0, 'm'},
                {"max-context", required_argument, 0, 'b'},
                {"max-recursion", required_argument, 0, 'r'},
                {"weight-cutoff", required_argument, 0, 'W'},
                {"time-cutoff", required_argument, 0, 't'},
                {"profile", no_argument, 0, 'p'},
                {0,0,0,0}
            };
        int option_index = 0;
        int c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT HFST_GETOPT_UNARY_SHORT "nxlwcdmb:r:W:t:p",
                            long_options, &option_index);
        if (-1 == c)
        {
            break;
        }


        switch (c)
        {
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-unary.h"
        case 'n':
            blankline_separated = false;
            break;
        case 'x':
            extract_patterns = on;
            break;
        case 'l':
            locate_mode = on;
            break;
        case 'w':
            print_weights = on;
            break;
        case 'c':
            count_patterns = on;
            break;
        case 'z':
            delete_patterns = on;
            break;
        case 'm':
            mark_patterns = off;
            break;
        case 'b':
            max_context = atoi(optarg);
            if (max_context < 0)
            {
                std::cerr << "Invalid argument for --max-context\n";
                return EXIT_FAILURE;
            }
            break;
        case 'r':
            max_recursion = atoi(optarg);
            if (max_recursion < 0)
            {
                std::cerr << "Invalid argument for --max-recursion\n";
                return EXIT_FAILURE;
            }
            break;
        case 'W':
            weight_cutoff = atof(optarg);
            if (weight_cutoff < 0.0)
            {
                std::cerr << "Invalid argument for --weight-cutoff\n";
                return EXIT_FAILURE;
            }
        case 't':
            time_cutoff = atof(optarg);
            if (time_cutoff < 0.0)
            {
                std::cerr << "Invalid argument for --time-cutoff\n";
                return EXIT_FAILURE;
            }
            break;
        case 'p':
            profile = true;
            break;
#include "inc/getopt-cases-error.h"
        }
        
        
        
    }
        // no more options, we should now be at the input filename
        if ( (optind + 1) < argc)
        {
            std::cerr << "More than one input file given\n";
            return EXIT_FAILURE;
        }
        else if ( (optind + 1) == argc)
        {
            if (inputfilename != NULL) {
                std::cerr << "More than one input file given\n";
                return EXIT_FAILURE;
            } else {
                inputfilename = hfst_strdup(argv[optind]);
                inputfile = hfst_fopen(inputfilename, "r");
                if (inputfile == stdin)
                {
                    free(inputfilename);
                    inputfilename = hfst_strdup("<stdin>");
                }
            }
            return EXIT_CONTINUE;
        } else
        {
            if (inputfilename == NULL) {
                std::cerr << "No input file given\n";
                return EXIT_FAILURE;
            } else {
                return EXIT_CONTINUE;
            }
        }


#include "inc/check-params-common.h"


    
    return EXIT_FAILURE;
}

int main(int argc, char ** argv)
{
    hfst_set_program_name(argv[0], "0.1", "HfstPmatch");
    hfst_setlocale();
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE) {
        return retval;
    }
#ifdef HAVE_READLINE
    // Disable tab completion
    rl_bind_key('\t', rl_insert);
#endif

    std::ifstream instream(inputfilename,
                           std::ifstream::binary);
    if (!instream.good()) {
        std::cerr << "Could not open file " << inputfilename << std::endl;
        return EXIT_FAILURE;
    }
    try {
        hfst_ol::PmatchContainer container(instream);
        container.set_verbose(verbose);
        if (extract_patterns != not_defined)
            container.set_extract_patterns(extract_patterns == on);
        if (locate_mode != not_defined)
            container.set_locate_mode(locate_mode == on);
        if (count_patterns != not_defined)
            container.set_count_patterns(count_patterns == on);
        if (delete_patterns != not_defined)
            container.set_delete_patterns(delete_patterns == on);
        if (mark_patterns != not_defined)
            container.set_mark_patterns(mark_patterns == on);
        if (max_context >= 0)
            container.set_max_context(max_context);
        if (max_recursion >= 0)
            container.set_max_recursion(max_recursion);
        container.set_profile(profile);
#ifdef _MSC_VER
        //hfst::print_output_to_console(true);
#endif
    return process_input(container, std::cout);
    } catch(HfstException & e) {
        std::cerr << "The archive in " << inputfilename << " doesn't look right."
            "\nDid you make it with hfst-pmatch2fst or make sure it's in weighted optimized-lookup format?\n";
        return 1;
    }
//     if (outfile != stdout) {
//         std::filebuf fb;
// fb.open(outfilename, std::ios::out);
// std::ostream outstream(&fb);
// return process_input(container, outstream);
// fb.close();
//     } else {

}
